docs/design/reflection-flow.md

```diff
@@ -0,0 +1,189 @@
+Activator.CreateInstance(t);
+// Linker should infer we need method Foo on the 2 types above
+var mi = t.GetMethod("Foo");
+// Linker should infer we need field Bar on the 2 types above
+var fi = t.GetMethod("Bar");
+```
+
+In an ideal world, this would be the extent of the reflection that can be made safe by the linker. Such small subset would not be practical because reflection often happens across method bodies. Instead of introducing cross-method dataflow analysis, we'll help linker reason about reflection happening across method bodies with annotations.
+
+## Cross-method annotations
+
+To document reflection use across methods, we'll introduce a new attribute `DynamicallyAccessedMembersAttribute` that can be attached to method parameters, the method return parameter, fields, and properties (whose type is `System.Type`, or `System.String`). The attribute will provide additional metadata related to linker-safety of the parameter or field.
+
+(When the attribute is applied to a location of type `System.String`, the assumption is that the string represents a fully qualified type name.)
+
+```csharp
+public sealed class DynamicallyAccessedMembersAttribute : Attribute
+{
+    public DynamicallyAccessedMembersAttribute(MemberKinds memberKinds)
+    {
+        MemberKinds = memberKinds;
+    }
+
+    public MemberKinds MemberKinds { get; }
+}
+
+[Flags]
+public enum MemberKinds
+{
+    DefaultConstructor = 1,
+    PublicConstructors = 3, // Maps to BindingFlags.Public
+    Constructors = 7, // Maps to BindingFlags.Public | BindingFlags.NonPublic
+    PublicProperties = 0x10,
+    Properties = 0x30,
+    PublicMethods = 0x40,
+    Methods = 0x70,
+    NestedTypes = 0x80,
+    // ...
+}
+```
+
+When a method or field is annotated with this attribute, two things will happen:
+* The method/field becomes potentially linker-unsafe. Linker will ensure that the values logically written to the annotated location (i.e. passed as a parameter, returned from the method, written to the field) can be statically reasoned about, or a warning will be generated.
+* The analysis of the value read from the annotated location will have richer information available and the linker can assume that linker-unsafe operations with an otherwise unknown value could still be safe.
+
+Example:
+
+```csharp
+class Program
+{
+    [DynamicallyAccessedMembers(MemberKinds.Methods)]
+    private static readonly Type _otherType;
+
+    static Program()
+    {
+        // _otherType is marked DynamicallyAccessedMembers - the linker
+        // ensures that the value written to this can be statically reasoned about
+        // and meets the MemberKinds restriction. So it will keep all methods on Foo.
+        _otherType = Type.GetType("Foo, FooAssembly");
+    }
+
+    static void Main()
+    {
+        // SAFE: _otherType was annotated as "some type that has all methods kept"
+        _otherType.GetMethod("RandomMethod");
```

Comment:
Can you elaborate on this example more or point to where the pattern is used in CoreLib so I can understand the need a little better?

On the surface this pattern could become always safe by putting the `GetType` and `GetMethod` together.

```
Type.GetType("Foo, FooAssembly").GetMethod("RandomMethod");
```

Reply:
Not sure which pattern in CoreLib had @MichalStrehovsky in mind, but `Lazy<T>` almost fits this as well. Some of the [`.ctors`](https://github.com/dotnet/runtime/blob/aa3870d606efb225e9f715993212ba3e5edd8098/src/libraries/System.Private.CoreLib/src/System/Lazy.cs#L208) which don't take a `valueCreator` effectively need the `T` to have a default `.ctor` (and be instantiatable). Internally it's implemented by eventually calling the [`Activator.CreateInstance`](https://github.com/dotnet/runtime/blob/aa3870d606efb225e9f715993212ba3e5edd8098/src/libraries/System.Private.CoreLib/src/System/Lazy.cs#L156) on it.

Now the way the code is currently written can't be directly annotated with the proposed attributes (since it actually doesn't store the `Type` instead it stores a `boolean` and relies on the `T`), but it could be refactored relatively easily to fit the pattern.

Reply:
I put `GetMethod`, this this one is more interesting with `Activator.CreateInstance`. E.g. here:

https://github.com/dotnet/runtime/blob/0a2f0e6da98a784601b37d19417b930a33b2c478/src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs#L610

Of course they could cache the `ConstructorInfo` instead, eliminating the need for the annotation.

But the building block (being able to annotate a field) is necessary to get custom attributes that have `System.Type` properties or constructor arguments working. I would like to make those work because they're pretty common.

Reply:
> it could be refactored relatively easily to fit the pattern.

Would this refactoring make `Lazy<T>` more expensive?

Reply:
Our linker would detect and correctly mark all of the reflection calls in `LicenseInteropProxy()` without the help of any annotations.  So those are OK.  

You are right that `object licContext = Activator.CreateInstance(_licInfoHelper)!;` is the tricky part.

Thinking out loud here.  If we detected the type that is assigned to the field, then we could store that knowledge.  Then when we come across the `Activator.CreateInstance` that having the field passed to it we could mark the ctor of all types that had been assigned to that field.  So I think we could actually automatically handle this scenario.

I guess what dancing around is I'm trying to figure out is the pros/cons of this proposal vs attributes such as `[RequiredMember]` that I've discussed here https://github.com/mono/linker/issues/797#issuecomment-549434815.  We are accumulating support for variety of these attributes to start trying them out.  I don't want to introduce things that would conflict with the wider .NET ecosystem but at the same time our use cases are slightly different which is why I'm trying to understand exactly what scenarios you are trying to solve.

Reply:
> I guess what I'm dancing around is I'm trying to figure out is the pros/cons of this proposal vs attributes such as `[RequiredMember]`

The main objective is be able to statically analyze whether an app is linker friendly. We want to avoid surprises (changes in semantics) at runtime. The goal is that if the linker issues no warnings, nothing should be broken by it.

The problem with annotations that just say "keep this" is that they don't have a connection to the specific reflection API usage within the method - they cannot be used in validating that the actual usage is safe. The linker either has to trust that the annotation on the method itself (e.g. `PreserveDependency`) fully covers what the method is doing, or worse yet, has to blindly trust that an annotation that is in a _different file_ will cover the pattern.

I'm not very happy about including `PreserveDependency` in the proposal for this reason too - it adds a blind spot to the linker and it's super easy to have code that is not in sync with the annotation, because the annotation duplicates what the code is doing. I keep pointing at https://github.com/dotnet/runtime/pull/2384.

This proposal is trying to come up with a way of doing this without duplicating the logic within the method, because that duplicated knowledge is hard to check for correctness statically.

Reply:
> Would this refactoring make Lazy<T> more expensive?
I don't think so - it's just a matter of storing for example `ConstructorInfo` instead of a boolean - so that there's something the linker can use to track the usage of the Type and the fact that it will be constructed.
Also this should only affect `Lazy<T>` without the `valueCreator` which seems to be pretty rare - for example I haven't seen it used in framework yet.

Reply:
ContructorInfo is the expensive reflection. It is much more expensive to create and deal with than Type.

Reply:
In that case we can store just the Type - it will seem redundant since the T is already there, but if we call it "typeToCreate" or something along those lines it should be OK.

Reply:
Storing Type would be better, but it still not clear how it would be done without introducing regression. `Lazy<T>` is perf critical e.g. for FSharp, and great deal of work went into making it as lean as possible.

Reply:
I think `Lazy<T>` would be solvable by one of two ways:

Mark the `Lazy<T>` constructor overload that makes it call into `CreateInstance<T>` as `LinkerUnsafe` & mark the `CreateViaDefaultConstructor()` as "linker shut up, this is safe".

Or,

We can implement the extension Vitek added as a TODO to the doc:

```csharp
class Lazy<[DynamicallyAccessedMembers(MemberKinds.DefaultConstructor)] T>
{
    // ...
}
```

This would ensure the `Activator.CreateInstance<T>` within lazy is not marked linker unsafe.

The drawback of the second approach is that now every use of `Lazy<T>` from generic code needs to be annotated because the annotations are contagious, even though the default constructor might not be needed if a factory delegate was provided:

```csharp
void Frob<T>()
{
    // This is now unsafe because the T is not marked [DynamicallyAccessedMembers(MemberKinds.DefaultConstructor)]
    // To fix it, we need to mark the T in Frob too
    Lazy<T> lazy = ...
}
```

Reply:
> The problem with annotations that just say "keep this" is that they don't have a connection to the specific reflection API usage within the method - they cannot be used in validating that the actual usage is safe. The linker either has to trust that the annotation on the method itself

Yes, I agree that is a down side.   We have users doing things that would fall into the "you should use source generators" category.  Many of them want something that is better than nothing which is where `[RequireMember]` style attributes come in.

I like the double checking that the connection allows, but I'm wondering if the fact that the connection can be made means that the linker can just auto detect these things.  @MichalStrehovsky  Is there a use case where the connection is maintained but the linker cannot auto detect the usage?  As I mentioned in my previous comment,  the example you pointed to could be automatically detected I believe.

Reply:
To some degree it's a tradeoff between complexity in the linker and explicit behavior in the API. For example, in some cases linker may be able to "propagate" these annotations to callers (over several jumps) - but then if one of the callers doesn't satisfy the necessary requirement (for example it passes in a Type which can't be figured out) the warning message would be relatively confusing as there would be no notion of it in the source code. The warning would say something like "The parameter `inputType` of `GetSomeData` requires a known type with all its fields" but the method `GetSomeData` would have nothing on it which would hint at that fact. Making the annotations explicit means that the behavior is to some degree self-documenting.

As for cases which we can't hope the linker to figure out: Casts (to and from `object` specifically), collections, and so on. In these cases the annotations could be used to hint the linker what's going on, but I don't think the linker could figure it out on its own.